codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/rustlang/RustModule.kt

```diff
@@ -5,23 +5,62 @@
 
 package software.amazon.smithy.rust.codegen.core.rustlang
 
-data class RustModule(val name: String, val rustMetadata: RustMetadata, val documentation: String? = null) {
-    fun render(writer: RustWriter) {
-        documentation?.let { docs -> writer.docs(docs) }
-        rustMetadata.render(writer)
-        writer.write("mod $name;")
+/**
+ * RustModule system.
+ *
+ * RustModules are idempotent, BUT, you must take care to always use the same module (matching docs, visibility, etc.):
+ * - There is no guarantee _which_ module will be rendered.
+ */
+sealed class RustModule {
+
+    /** lib.rs */
+    object LibRs : RustModule()
+
+    /**
+     * LeafModule
+     *
+     * A LeafModule is _all_ modules that are not `lib.rs`. To create a nested leaf module, set `parent` to a module
+     * _other_ than `LibRs`.
+     *
+     * To avoid infinite loops, avoid setting parent to itself ;-)
+     */
+    data class LeafModule(
+        val name: String,
+        val rustMetadata: RustMetadata,
+        val documentation: String? = null,
+        val parent: RustModule = LibRs,
+        val inline: Boolean = false,
+    ) : RustModule() {
+        init {
+            check(!name.contains("::")) {
```

Comment:
Should we also run the `name` through `RustReservedWords.escapeIfNeeded()`?

The server creates modules with names based on shape names.

Reply:
yeah that sounds like a good idea to me!